cdc/sink/codec/avro.go

```diff
@@ -160,30 +152,86 @@ func (a *AvroEventBatchEncoder) Size() int {
 	return sum
 }
 
-func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion uint64, cols []*model.Column, colInfos []rowcodec.ColInfo, tz *time.Location) (*avroEncodeResult, error) {
+func (a *AvroEventBatchEncoder) avroEncode(
+	ctx context.Context,
+	e *model.RowChangedEvent,
+	topic string,
+	key bool,
+) (*avroEncodeResult, error) {
+	var cols []*model.Column
+	var colInfos []rowcodec.ColInfo
+	var enableTiDBExtension bool
+	var schemaManager *AvroSchemaManager
+	var operation string
+	if key {
+		cols, colInfos = e.HandleKeyColInfos()
+		enableTiDBExtension = false
+		schemaManager = a.keySchemaManager
+	} else {
+		cols = e.Columns
+		colInfos = e.ColInfos
+		enableTiDBExtension = a.enableTiDBExtension
+		schemaManager = a.valueSchemaManager
+		if e.IsInsert() {
+			operation = "c"
+		} else if e.IsUpdate() {
+			operation = "u"
+		} else {
+			log.Error("unknown operation", zap.Reflect("rowChangedEvent", e))
+			return nil, cerror.ErrAvroEncodeFailed.GenWithStack("unknown operation")
+		}
+	}
+
+	namespace := getAvroNamespace(a.namespace, e.Table)
+
 	schemaGen := func() (string, error) {
-		schema, err := ColumnInfoToAvroSchema(table.Table, cols)
+		schema, err := rowToAvroSchema(
+			namespace,
+			e.Table.Table,
+			cols,
+			colInfos,
+			enableTiDBExtension,
+			a.decimalHandlingMode,
+			a.bigintUnsignedHandlingMode,
+		)
 		if err != nil {
 			return "", errors.Annotate(err, "AvroEventBatchEncoder: generating schema failed")
 		}
 		return schema, nil
 	}
 
-	// TODO pass ctx from the upper function. Need to modify the EventBatchEncoder interface.
-	avroCodec, registryID, err := manager.GetCachedOrRegister(context.Background(), *table, tableVersion, schemaGen)
+	avroCodec, registryID, err := schemaManager.GetCachedOrRegister(
+		ctx,
+		topic,
+		e.TableInfoVersion,
+		schemaGen,
+	)
 	if err != nil {
 		return nil, errors.Annotate(err, "AvroEventBatchEncoder: get-or-register failed")
 	}
 
-	native, err := rowToAvroNativeData(cols, colInfos, tz)
+	native, err := rowToAvroData(
+		cols,
+		colInfos,
+		e.CommitTs,
+		operation,
+		enableTiDBExtension,
+		a.decimalHandlingMode,
+		a.bigintUnsignedHandlingMode,
+	)
 	if err != nil {
 		return nil, errors.Annotate(err, "AvroEventBatchEncoder: converting to native failed")
 	}
 
 	bin, err := avroCodec.BinaryFromNative(nil, native)
 	if err != nil {
 		return nil, errors.Annotate(
-			cerror.WrapError(cerror.ErrAvroEncodeToBinary, err), "AvroEventBatchEncoder: converting to Avro binary failed")
+			cerror.WrapError(
+				cerror.ErrAvroEncodeToBinary,
```

Comment:
`errors.Annotate` is verbose here, just `errors.Trace(err)` should be ok. add a log to help debug the problem if necessary.